edx/frontend-app-library-authoring
# [BD-14] [SE-2937] Authoring page

This PR introduces the new 'Authoring Page' which is intended to replace the current Library Detail view (pending a follow-up PR to be made by @arbrandes containing complex library capabilities).

The PR has a few tricks up its sleeve-- the first is that it generalizes the block editing slice to allow for multiple blocks to be tracked, thus permitting us to view the previews of all the blocks in a library at once.

The second is that it handles showing and hiding of previews, and saving the user's preference to localStorage.

The third is that it refines some of the testing tools that were made previously to handle cases that came out of testing this PR. One new feature is the ability to spy on dispatch calls, and another is mocks for local storage. The testing environment now better mirrors real world use, allowing for greater fidelity in tests.

One more thing this new page offers is that it is rewritten using Paragon and does not share its markup with the studio. This has allowed much more succinct templates.

**Discussions**: https://docs.google.com/document/d/1hRrSmWWlGHhi-bh9AIlpSuLtJFnn2GpavNe8OQxthbE/

**Dependencies**: None

**Screenshots**:

![image](https://user-images.githubusercontent.com/1099483/95004455-cefcf780-05b0-11eb-817c-1c7643ad5a8c.png)

![image](https://user-images.githubusercontent.com/1099483/95004462-ea680280-05b0-11eb-93b0-6fd9fbe1e992.png)

WIth previews hidden:

![image](https://user-images.githubusercontent.com/1099483/95004469-05d30d80-05b1-11eb-8402-b9131dbabd3d.png)


**Merge deadline**: Sooner is better than later.

**Testing instructions**:

1. Create a new Video/Problem library, or visit an existing one.
2. Use the 'Add Component' button, which should be renamed 'Add Video' or 'Add Problem' depending on the library type you're working with. Verify a new block has been created and that its preview loads. (It may be slow in a dev environment).
3. Add another component.
4. Toggle show and hide previews. Verify that the setting is saved on refresh.
5. Delete components.
6. Visit a complex library and verify it goes to the old LibraryPage instead of the LibraryAuthoringPage for now.
7. Publish your changes.
8. Change something.
9. Revert your changes. (It may take a moment)

**Author notes and concerns**:

1. The studio has a 'copy' button which is not implemented here. I attempted to implement it, but could not get copying to work using the existing endpoint used by old libraries, nor did I see an obvious endpoint in the new API for doing so. I imagine we'll want this feature in the future, so unless I've missed something, this will probably need to be a follow-up task.
2. The 'edit' button exists but can't be rendered currently because it only displays for blocks that can be edited, and neither the Problem nor the Video blocks can be.
3. As an addendum to 2, the editing functionality would currently go to the existing block edit page anyway-- I didn't flesh it out since we can't use it at present. Having settings appear in a modal will need to be a future task.

**Reviewers**
